File: byke2d/radix/radixsort.go
package radix

import "math"

//go:inline
func floatToSortableU32(f float32) uint32 {
	x := math.Float32bits(f)
	mask := uint32(int32(x)>>31) | 0x80000000
	return x ^ mask
}

func radixsortGo(values []Value, scratch []Value) {
	src := values
	dst := scratch

	const RADIX = 256

	var count [RADIX]uint32

	for pass := 0; pass < 4; pass++ {
		clear(count[:])

		shift := uint(pass * 8)

		/* Histogram */
		for i := uint32(0); i < uint32(len(src)); i++ {
			v := src[i] // bounds-checked, but usually eliminated
			k := floatToSortableU32(v.Key)
			count[(k>>shift)&0xFF]++
		}

		/* Prefix sum */
		sum := uint32(0)
		for i := 0; i < RADIX; i++ {
			c := count[i]
			count[i] = sum
			sum += c
		}

		/* Scatter */
		for i := uint32(0); i < uint32(len(src)); i++ {
			v := src[i]
			k := floatToSortableU32(v.Key)

			idx := (k >> shift) & 0xFF

			// local alias helps compiler avoid repeated bounds checks
			d := dst
			c := count[idx]
			d[c] = v
			count[idx] = c + 1
		}

		// swap buffers
		src, dst = dst, src
	}

	if &src[0] != &values[0] {
		panic("invariant: even number of sort passes must lead to src == values")
	}
}

File: byke2d/radix/radix_c/radixsort.c
#include "radixsort.h"

#include <string.h>

/* Convert float bit-pattern into sortable uint32_t. */
static inline uint32_t float_to_sortable_u32(float f) {
  union {
    float f;
    uint32_t u;
  } v = {f};
  uint32_t x = v.u;

  uint32_t mask = ((int32_t)x >> 31) | 0x80000000u;
  return x ^ mask;
}

void radixsort_c(value_t *restrict data, value_t *restrict scratch,
                 uint32_t n) {
  value_t *restrict src = data;
  value_t *restrict dst = scratch;

  const uint32_t RADIX = 256;

  for (int pass = 0; pass < 4; ++pass) {
    uint32_t count[RADIX] = {};

    const int shift = pass * 8;

    /* Histogram */
    for (uint32_t i = 0; i < n; ++i) {
      uint32_t k = float_to_sortable_u32(src[i].key);
      count[(k >> shift) & 0xFF]++;
    }

    /* Prefix sum */
    uint32_t sum = 0;
    for (uint32_t i = 0; i < RADIX; ++i) {
      uint32_t c = count[i];
      count[i] = sum;
      sum += c;
    }

    /* Scatter */
    for (uint32_t i = 0; i < n; ++i) {
      value_t v = src[i];
      uint32_t k = float_to_sortable_u32(v.key);

      uint32_t idx = (k >> shift) & 0xFF;
      dst[count[idx]++] = v;
    }

    /* swap buffers */
    value_t *tmp = src;
    src = dst;
    dst = tmp;
  }
}
